### PR TITLE
Add `File#delete?` and `Dir#delete?`

### PR DESCRIPTION
**Pull request by didactic-drunk**
Add `File.delete?(path) => Bool` indicating success

There are 2 common patterns for conditionally deleting files/dirs.  Both of which have problems.
* `File.delete(path) if File.exists?(path)` # Unnecessary stat call and race condition
* `File.delete(path) rescue nil` # Rescues more than necessary.  May hide subtle bugs

Also DRY.

These patterns are used > 10 times in crystal source and very common elsewhere.

### CODE REVIEW

**Comment by j8r** [comment]
It will be more efficient to also implement and use `Crystal::System::File.delete?`/`Crystal::System::Dir.delete?`

**Comment by didactic-drunk** [comment]
> It will be more efficient to also implement and use `Crystal::System::File.delete?`/`Crystal::System::Dir.delete?`

Do you mean `Crystal::System::File.delete` -> `Crystal::System::File.delete?` so the `Exception` is never raised when `File.delete?` is called?

File: src/dir.cr
Comment (j8r) [inline src/dir.cr:286]: The same can be done here too to avoid raising in the first place.

Also, another option instead of a `raise_on_missing` argument is to do something like:
```cr
def delete(path) : Nil
  delete?(path) || raise "..."
end
```
For me it is more idiomatic, this kind of logic is already present in the stdlib.

Comment (didactic-drunk) [reply src/dir.cr:286]: Ok,but then I need to duplicate the exception string *exactly* or the tests fail.

Comment (j8r) [reply src/dir.cr:286]: I meant that it will be done in `Crystal::System::Dir.delete?(path) : Bool`, which won't raise any exception.
There is only one error possible so it can be done.

Comment (didactic-drunk) [reply src/dir.cr:286]: This can't be solved with a `System::Dir::delete?` returning `Bool`.  `delete?` returns `false` for `ENOENT` and raises for all other errors.

I suppose `System...delete?` could return `Errno`?

Comment (j8r) [reply src/dir.cr:286]: I believe it is fine to make it return `Bool` – `true` when the directory/file is present `false`, otherwise – and keep it exceptionnaly raise for other, less likely errors.

Comment (j8r) [reply src/dir.cr:286]: That's also how you implemented `delete?` in `Dir` and `File` at the top-level at first: only rescue `File::NotFoundError` exceptions.

**Comment by asterite** [comment]
Please, please, stop discussing this here how it should work and how it should be implemented. Without input from a core member changes to PRs shouldn't be made. Otherwise it's a waste of time for the PR author. Not to say that non core members comments shouldn't be heard, but one shouldn't immediately act upon them.

**Comment by asterite** [comment]
That is, if you don't want to lose time on endless discussion without having a clear agreement on what we all want to do.

File: src/dir.cr
Comment (beta-ziliani) [reply src/dir.cr:286]: I like what this PR brings and what @j8r said: it makes sense to port the same approach to `system/*/dir.cr` and code something like:

```crystal
def self.delete(path, *, raise_on_missing : Bool) : Bool
    err = LibC.rmdir(path.check_no_null_byte)

    if err != -1
      true
    elsif !raise_on_missing && Errno.value == Errno::ENOENT
      false
    else
      raise ::File::Error.from_errno("Error deleting dir", file: path)
    end
end
```

⚠️ I'm not paying attention to details!

What do you think @didactic-drunk ?

Comment (didactic-drunk) [reply src/dir.cr:286]: Sure.  @beta-ziliani, you're new to me.  This will be approved or still needs @asterite approval?

Comment (beta-ziliani) [reply src/dir.cr:286]: Nice to meet you! You'll find me in the [team section](https://www.crystal-lang.org/team) ;-)
Any PR needs *two* approvals from core team members (see the aforementioned link). I think this is close to get the approvals. Personally, I'd like to see the change proposed above, but I can take care of that if you prefer.

Comment (j8r) [reply src/dir.cr:286]: Is it preferable to have an `raise_on_missing` argument or have two methods instead, one that can return nil and the other calling it and raising on nil. A bit like `#to_i?` and `#to_i`, or `Enumerable#first?` and `Enumerable#first` (raising).

Comment (asterite) [reply src/dir.cr:286]: A branch that potentially raises, even if that's impossible, is technically worse that code that can never raise.

Comment (beta-ziliani) [reply src/dir.cr:286]: From an API standpoint, it doesn't matter as this is about the internals. From a technical point, it's an optimization that could be done later. It's already making a difference IMO.

Comment (asterite) [reply src/dir.cr:286]: Right. I think what's important in this PR is to have something working. The performance differences or styles will probably not matter unless you are deleting files in a hot loop, which doesn't sound like something usual.

Comment (didactic-drunk) [reply src/dir.cr:288]: @j8r I agree 1 method without an argument is a better approach, however the specs require a specific error message to pass.  Either raise is duplicated exactly and kept in sync should the message ever change or an argument is used.  Or the specs are changed.  Or errors are reworked somehow.

My approach was I think the least invasive and I hope fastest to get approved.  Past experience has shown my more complicated PR's stagnate for **very** long periods.

My preference is:
1) Get this merged so I can begin using it
2) Let someone else optimize it later and take the burden of keeping error messages in sync or coming up with another solution

File: src/dir.cr
Comment (straight-shoota) [inline src/dir.cr:284]:
```suggestion
  # Returns `false` if the directory does not exist.
```

File: src/file.cr
Comment (straight-shoota) [inline src/file.cr:352]:
```suggestion
  # Deletes the file at *path*.
  # Returns `false` if the file does not exist.
```

**Review by straight-shoota — APPROVED** [review]
I just have a tiny proposal for improving the documentation. Everything else looks good.

**Comment by beta-ziliani** [comment]
WASI is failing now, `./system/wasi/dir.cr` needs the same change. Let me know @didactic-drunk if you want me to take it. And with that this should be :ok_hand:

**Comment by didactic-drunk** [comment]
> WASI is failing now, `./system/wasi/dir.cr` needs the same change. Let me know @didactic-drunk if you want me to take it. And with that this should be ok_hand

@beta-ziliani  If the changes were insufficient please take.

**Review by beta-ziliani — APPROVED** [review]

**Pull request merged**
